README.md
[![Build Status](https://travis-ci.com/macornwell/MusksUltimateMessageSafe.svg?branch=master)](https://travis-ci.com/macornwell/MusksUltimateMessageSafe)
[![Coverage Status](https://coveralls.io/repos/github/macornwell/MusksUltimateMessageSafe/badge.svg?branch=master)](https://coveralls.io/github/macornwell/MusksUltimateMessageSafe?branch=master)
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)



# Crack Elon Musk's Ultimate Message Safe

## ???
Why is Elon Musk so successful? 

How has this person been able to generate momentum across so many technological areas across the planet, seemingly so
easily? If you have watched every Elon Musk presentation you will find out quickly that he isn't a great speaker. It
isn't even clear that he is even all that inspiring and yet he has generated a spark that has caused even the largest
most powerful government ever created to become significantly more interested in space.

Elon Musk may have created a blackhole of human attention.

## OK, but what is it about Elon Musk?
After considerable thinking, I may have come to the realization why he is in-fact as successful as he is. The reality is
what he has, every succesful person has in common and it is what is driving them to do unbelievably large things.

Because this statement has value and is worth trying to obtain, I decided to create a safe for safe keeping. A digital
safe that requires a password to enter.

### How To Open the Safe
1. Select the starting number 'A'.
2. Turn the safe to the left 'X' number of times. This will result in a galactically large number.
3. Turn the safe to the right 'Y' number of times. This will result in a small number, at least for us on earth.
4. Finally, turn the safe to the left 'Z' number of times to arrive at the final message.

If you take a look at the source code you will see a number of different language iterations. Choose which one that you
are comfortable with. You will notice that the operations of the safe, work just like a regular safe.
I have encoded the message in this application which is written across a number of programming languages. Choose
whichever one suits you best and try to crack the safe. My recommendation is to read the source code and play around with the applications that
are provided. 

### Languages Supported
This application is written in the following languages.
- C++
- Python
- Haskell
- JavaScript

### Dependencies
Because exact calculations were needed on very large integers (in excess of 10^128) external libraries were needed.
- The C++ version requires gmp and zlib.
- The Python version requires sympy.
- The Haskell version requires the Haskell stack.
- The JavaScript version requires mocha, pako, and big-integer.

## FAQ
**1. Are there any hints? I can't figure it out.**
*Keep trying. You can do it, you do not need any hints.*

**2. I know English well, but I don't know any of the languages you have.**
*While I'd like to support a language you understand (email me), perhaps its time for you to learn one 
that I understand.*

**3. Couldn't I just ask around what the safe combination is?**
*You could but you wouldn't learn anything, and you wouldn't have had the experience of cracking this safe.*


haskell/ChangeLog.md
# Changelog for MusksUltimateMessageSafe

## Unreleased changes


python/tests.py
"""
Copyright (C) 2018 Michael Cornwell

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
from unittest import TestCase
import sympy
from musks_ultimate_message_safe import MusksUltimateMessageSafe


class MusksUltimateMessageSafeTest(TestCase):

    def setUp(self):
        pass

    def test_turn_ultimate_dial_left_to_get_a_big_number(self):
        safe = MusksUltimateMessageSafe()
        starting_number = 1
        rotations_left = 1
        big_integer = safe.turn_ultimate_dial_left_to_get_a_big_number(starting_number, rotations_left)
        expected_result = sympy.Integer("226169770062649671975095064650885165073473873466476969629424020831684677671512"
                                        "831015195781658332172748345988614697960374030307145316544864031281391887096765"
                                        "740000987048034")
        self.assertEqual(big_integer, expected_result)

    def test_turn_utlimate_dial_right_to_get_a_small_number(self):
        safe = MusksUltimateMessageSafe()
        rotations_right = 221
        big_number = sympy.Integer("28995220320916676973132026074599907928106961378772759180718885321628947773482461446"
                                   "70540447794367184535471847894388467904654907613165448640312813918870967657356029405"
                                   "36930")
        smaller_number = safe.turn_ultimate_dial_right_to_get_a_small_number(rotations_right, big_number)
        expected_result = sympy.Integer("270862412836236458252642268867205902245058122461237727136720590227132772636859"
                                        "002293279263292374615627102836529223686157599858806364179862406156652631202476"
                                        "237223526325261823741786229064802786300663426740")
        self.assertEqual(str(smaller_number), str(expected_result))

    def test_turn_ultimate_dial_left_until_the_message_appears(self):
        safe = MusksUltimateMessageSafe()
        rotations_left = 10
        smaller_number = sympy.Integer("13258259824895728903745280937498234")
        message = safe.turn_ultimate_dial_left_until_the_message_appears(rotations_left, smaller_number)
        expected_message = 'RȃȂɕȫǑǹǔ\x0e'
        self.assertEqual(message, expected_message)



javascript/test.js
/*
 Copyright (C) 2018 Michael Cornwell

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var assert = require('assert');
var bigInt = require('big-integer');
var muskSafe = require('./musks-ultimate-message-safe.js');

describe('MusksUltimateMessageSafe', function() {
  describe('#turnUltimateDialLeftToGetABigNumber()', function() {
    it('should handle starting_value of 1 and rotations_left of 1', function() {
      var safe = new muskSafe();
      var starting_value = 1;
      var rotations_left = 1;
      var big_int = safe.turnUltimateDialLeftToGetABigNumber(starting_value, rotations_left);
      var expected = "2.2616977006264967e+170";
      var result = String(big_int);
      assert.equal(result, expected);
    });
  });

  describe('#turnUltimateDialRightToGetASmallNumber', function() {
    it('Should be able to handle rotations right 221 and provided big int', function() {
      var safe = new muskSafe();
      var rotations_right = 221;
      var big_number = bigInt("28995220320916676973132026074599907928106961378772759180718885321628947773482461446" +
        "70540447794367184535471847894388467904654907613165448640312813918870967657356029405" +
        "36930");
      var smaller_number = safe.turnUltimateDialRightToGetASmallNumber(rotations_right, big_number);
      var expected_result = bigInt("270862412836236458252642268867205902245058122461237727136720590227132772636859" +
        "002293279263292374615627102836529223686157599858806364179862406156652631202476" +
        "237223526325261823741786229064802786300663426740");
      assert.equal(String(smaller_number), String(expected_result));
    });
  });


  describe('#turnUltimateDialLeftUntilMessageAppears', function() {
    it('Should be able to handle rotations 10 and provided smaller_number', function() {
      var safe = new muskSafe();
      var rotations_left = 10;
      var smaller_number = bigInt("13258259824895728903745280937498234");
      var message = safe.turnUltimateDialLeftUntilTheMessageAppears(rotations_left, smaller_number);
      var expected_message = 'RȃȂɕȫǑǹǔ\x0e';
      assert.equal(message, expected_message);
    });
  });
});


haskell/test/Spec.hs
module Main where

import Test.Hspec
import Test.QuickCheck
import Control.Exception (evaluate)
import qualified Lib

main :: IO ()
main = hspec $ do
  describe "Lib.convertNumberToLetters" $ do
    it "returns '' for 1" $ Lib._convertNumberToLetters 1 `shouldBe` ""
    it "returns '' for 2" $ Lib._convertNumberToLetters 2 `shouldBe` ""
    it "returns 'd' for 3" $ Lib._convertNumberToLetters 3 `shouldBe` "d"
    it "returns 'e' for 4" $ Lib._convertNumberToLetters 4 `shouldBe` "e"
    it "returns 'j' for 9" $ Lib._convertNumberToLetters 9 `shouldBe` "j"
    it "returns 'k' for 10" $ Lib._convertNumberToLetters 10 `shouldBe` "k"
    it "returns 'l' for 11" $ Lib._convertNumberToLetters 11 `shouldBe` "l"
    it "returns 'fmdihge' for 51238764" $ Lib._convertNumberToLetters 51238764 `shouldBe` "fmdihge"

  describe "Lib.unScrunchBinary" $ do
    it "returns '[1]' for 2" $ Lib._unScrunchBinary 2 `shouldBe` [1]
    it "returns '[3]' for 3" $ Lib._unScrunchBinary 3 `shouldBe` [3]
    it "returns '[1,1]' for 4" $ Lib._unScrunchBinary 4 `shouldBe` [1,1]
    it "returns '[1,3]' for 5" $ Lib._unScrunchBinary 5 `shouldBe` [1,3]
    it "returns '[1,3,1]' for 10" $ Lib._unScrunchBinary 10 `shouldBe` [1,3,1]
    it "returns '[3, 3, 3, 1, 1, 1, 1, 1, 3, 3, 3, 3, 1, 1, 3, 3]' for 123123" $ Lib._unScrunchBinary 123123 `shouldBe` [3, 3, 3, 1, 1, 1, 1, 1, 3, 3, 3, 3, 1, 1, 3, 3]
    it "returns '[1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1]' for 2261697700626496" $ Lib._unScrunchBinary 2261697700626496 `shouldBe` [1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1]


  describe "Lib.consultBlackBallWithLetter" $ do
    it "returns '0' for 'a'" $ Lib._consultBlackBallWithLetter 'a' `shouldBe` 0
    it "returns '8' for 'b'" $ Lib._consultBlackBallWithLetter 'b' `shouldBe` 8
    it "returns '10' for 'c'" $ Lib._consultBlackBallWithLetter 'c' `shouldBe` 10
    it "returns '38' for 'u'" $ Lib._consultBlackBallWithLetter 'u' `shouldBe` 38
    it "returns '40' for 'v'" $ Lib._consultBlackBallWithLetter 'v' `shouldBe` 40
    it "returns '0' for 'w'" $ Lib._consultBlackBallWithLetter 'w' `shouldBe` 0
    it "returns '0' for 'y'" $ Lib._consultBlackBallWithLetter 'y' `shouldBe` 0
    it "returns '0' for 'z'" $ Lib._consultBlackBallWithLetter 'z' `shouldBe` 0

  describe "Lib.consultBlackBallWithNumber" $ do
    it "returns 'a' for '0'" $ Lib._consultBlackBallWithNumber 0 `shouldBe` 'a'
    it "returns 'b' for '1'" $ Lib._consultBlackBallWithNumber 1 `shouldBe` 'b'
    it "returns 'c' for '2'" $ Lib._consultBlackBallWithNumber 2 `shouldBe` 'c'
    it "returns 'i' for '8'" $ Lib._consultBlackBallWithNumber 8 `shouldBe` 'i'
    it "returns 'k' for '10'" $ Lib._consultBlackBallWithNumber 10 `shouldBe` 'k'
    it "returns '0' for '40'" $ Lib._consultBlackBallWithNumber 40 `shouldBe` 'a'
    it "returns 'u' for '20'" $ Lib._consultBlackBallWithNumber 20 `shouldBe` 'u'
    it "returns 'v' for '21'" $ Lib._consultBlackBallWithNumber 21 `shouldBe` 'v'
    it "returns 'b' for '22'" $ Lib._consultBlackBallWithNumber 22 `shouldBe` 'b'
    it "returns 'c' for '23'" $ Lib._consultBlackBallWithNumber 23 `shouldBe` 'c'

  describe "Lib.unpackToOffsetList" $ do
    it "returns '[20, 16, 0, 24, 24, 16, 32, 16, 0, 22]' for 'gfaiifqfah'" $ Lib._unpackToOffsetList "gfaiifqfah" `shouldBe` [20, 16, 0, 24, 24, 16, 32, 16, 0, 22]
    it "returns '[0]' for 'a'" $ Lib._unpackToOffsetList "a" `shouldBe` [0]
    it "returns '[8]' for 'b'" $ Lib._unpackToOffsetList "b" `shouldBe` [8]
    it "returns '[10]' for 'c'" $ Lib._unpackToOffsetList "c" `shouldBe` [10]
    it "returns '[10]' for 'c'" $ Lib._unpackToOffsetList "c" `shouldBe` [10]





cpp/src/main.cpp
/*
Copyright (C) 2018 Michael Cornwell

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <iostream>
#include <chrono>
#include <thread>
#include <string>
#include "gmp.h"
#include "musks_ultimate_message_safe.h"

const int DRAMATIC_WAIT_IN_SECONDS = 1;

void StartSafeProgram() {
    std::cout << "Welcome to Elon Musk's Ultimate Message Safe. Only Elon may enter." << std::endl
              << "The safe requires the following four (4) inputs:" << std::endl
              << std::endl
              << "    [STARTING_NUMBER]     The number to start the safe on." << std::endl
              << "    [TURNS_LEFT]          The number of turns left to turn the dial." << std::endl
              << "    [TURNS_RIGHT]         The number of turns right to turn the dial." << std::endl
              << "    [TURNS_LEFT]          The number of turns left to turn the dial to get the message." << std::endl
              << std::endl;

    std::cout << "What number should the safe start on?" << std::endl;
    unsigned long starting_number = 0;
    std::cin >> starting_number;
    std::cout << "How many turns left?" << std::endl;
    unsigned long turns_left = 0;
    std::cin >> turns_left;
    std::cout << "How many turns right?" << std::endl;
    unsigned long turns_right = 0;
    std::cin >> turns_right;
    std::cout << "How many turns left?" << std::endl;
    unsigned long turns_left_2 = 0;
    std::cin >> turns_left_2;

    MusksUltimateMessageSafe safe;

    std::string message;
    mpz_t big_number = {};
    mpz_init(big_number);
    mpz_set_ui(big_number, 0);
    mpz_t smaller_number = {};
    mpz_init(smaller_number);
    mpz_set_ui(smaller_number, 0);

    std::cout << "You set the safe on number( " << starting_number <<  " )." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));
    std::cout << "You start to turn( " << turns_left << " )turns to the left, slowly." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));

    safe.TurnUltimateDialLeftToGetABigNumber(starting_number, turns_left,big_number);

    std::cout << "You lose grip on the dial and it continues turning on its own." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));
    std::unique_ptr<char[]> large_number_pointer(mpz_get_str(NULL, 10, big_number));
    std::string large_number(&large_number_pointer[0]);
    std::cout << "Something kicks in and boosts the spinning of the combination lock. The safe spins out of control to "
              << large_number << "." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));

    safe.TurnUtlimateDialRightToGetASmallNumber(turns_right, big_number, smaller_number);

    std::unique_ptr<char[]> small_number_pointer(mpz_get_str(NULL, 10, smaller_number));
    std::string small_number_string(&small_number_pointer[0]);
    std::cout << "The safe turns( " << turns_right << " )rotations to the right and settles on "
              << small_number_string << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));

    message = safe.TurnUltimateDialLeftUntilTheMessageAppears(turns_left_2, smaller_number);

    std::cout << "You turn the safe left (" << std::to_string(turns_left_2)
              <<") times and the following message appears...  ";
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));
    std::cout << std::endl << std::endl << "\"" << message << "\"" << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(DRAMATIC_WAIT_IN_SECONDS));
}

int main(int argc, char *argv[]) {
    StartSafeProgram();
    return 0;
}

haskell/src/Lib.hs
module Lib
  where

import qualified Data.Map.Strict as Map
import qualified Data.ByteString.Lazy.Char8 as BC
import qualified Codec.Compression.Zlib as Zlib
import Data.Maybe
import Data.Bits
import Data.Char

bitesShifted = 0x4 :: Int
ultimateAnswer = 42
high = 0x1
luckyNumber = 0x3
courseCorrection = 320
bassDecimal = 10 :: Int
twentyTwoSidedDieEncasedInWater = [
  0, 8, 10, 12, 14, 16, 20, 22, 24, 25, 26, 27, 28,
  29, 30, 31, 32, 33, 34, 36, 38, 40]
letters = ['a'..'z']
lettersIndex = [0..(length letters - 1)]
indexToLetter = Map.fromList [(x, letters !! x)| x <- lettersIndex]
letterToIndex = Map.fromList [(letters !! x ,x)| x <- lettersIndex]
overHexed = 0xF + 0x01
ultimateWindageAdjustment = ultimateAnswer + 32 :: Int
secretSeed = BC.pack "\x78\x9c\x0d\x8e\x81\x09\x00\x40\x08\x02\x57\xb2\xcf\xb4\xf6\x5f\xec\x45\x48\x0c\x2e\x7b\x4f\xa5\xb3\x01\x3d\xf1\xe4\x3a\x0f\x6e\x20\x6a\xb0\x3b\x15\x73\xd3\xbd\x99\x12\xad\x8b\xde\xf1\x11\x0f\xdb\xa5\xa5\xec\x80\x53\x2f\x11\xf1\x1b\x6f\xb8\xed\x7e\xe5\x67\x6e\x73\x6e\x57\xc5\x54\x9d\xd0\x26\x3a\x72\x71\x72\x60\xc8\x55\x36\xe1\xab\xaf\x76\x8d\x3c\xa2\x71\x8f\x90\x26\x4c\xeb\x1a\x1f\xd2\x9b\x23\x21"

_decryptSecretSeed :: Integer
_decryptSecretSeed = read $ BC.unpack (Zlib.decompress secretSeed) :: Integer

_continueTurningLeftUntilClickIsHeard :: [Int] -> [Int] -> [Int]
_continueTurningLeftUntilClickIsHeard wholeValues offsets = [
  wholeValues !! i * ultimateAnswer + (offsets !! i)
    | i <- [0..(length wholeValues - 1)]
  ]

_consultBlackBallWithNumber :: Int -> Char
_consultBlackBallWithNumber x
  | x == size = fromMaybe 'a' (Map.lookup 1 indexToLetter)
  | x == size + 1 = fromMaybe 'a' (Map.lookup 2 indexToLetter)
  | x >= length indexToLetter = fromMaybe 'a' (Map.lookup 0 indexToLetter)
  | otherwise = fromMaybe 'a' (Map.lookup x indexToLetter)
  where size = length twentyTwoSidedDieEncasedInWater

_consultBlackBallWithLetter :: Char -> Int
_consultBlackBallWithLetter x =
  case Map.lookup x letterToIndex of
    Nothing -> 0
    Just maybeIndex ->
      if maybeIndex > length twentyTwoSidedDieEncasedInWater - 1
        then
        0
      else
        twentyTwoSidedDieEncasedInWater !! maybeIndex

_unScrunchBinary :: Int -> [Int]
_unScrunchBinary scrunched =
  if scrunched > 1
    then
      _unScrunchBinary (scrunched `shiftR` 1) ++
      [case (.&.) scrunched 1 of
        1 -> luckyNumber
        0 -> high]
  else
    []

_recursiveConverter :: String -> String
_recursiveConverter [] = ""
_recursiveConverter [x] = if x == '1' || x == '2' then "" else _consultBlackBallWithNumber (digitToInt x) : ""
_recursiveConverter (x:xs) = do
  let next = head xs
  if x == '1' || x == '2'
  then
    _consultBlackBallWithNumber (read (x : [next]) :: Int) : _recursiveConverter (tail xs)
  else
    _consultBlackBallWithNumber (digitToInt x) : _recursiveConverter xs

_convertNumberToLetters :: Integer -> String
_convertNumberToLetters x = do
  let stringNumber = show x
  _recursiveConverter stringNumber

_unpackToOffsetList :: String -> [Int]
_unpackToOffsetList x = [_consultBlackBallWithLetter y | y <- x]

turnUltimateDialLeftToGetABigNumber :: Integer -> Integer -> Integer
turnUltimateDialLeftToGetABigNumber startingNumber rotationsLeft =
  (((startingNumber + rotationsLeft) * rotationsLeft) ^ ultimateAnswer) + _decryptSecretSeed

turnUltimateDialRightToGetASmallNumber :: Integer -> Integer -> Integer
turnUltimateDialRightToGetASmallNumber rotationsRight bigNumber = do
  let scrunched = take overHexed (show bigNumber)
  let packed = drop overHexed (show bigNumber)
  let packed1 = take ultimateWindageAdjustment packed
  let packed2 = drop ultimateWindageAdjustment packed
  let packed1Letters = _convertNumberToLetters (read packed1 :: Integer)
  let packed2Letters = _convertNumberToLetters (read packed2 :: Integer)
  let offset1 = _unpackToOffsetList packed1Letters
  let offset2 = _unpackToOffsetList packed2Letters
  let wholeValues = _unScrunchBinary (read scrunched :: Int)
  let wholeValues1 = _continueTurningLeftUntilClickIsHeard wholeValues offset2
  let wholeValues2 = _continueTurningLeftUntilClickIsHeard wholeValues1 offset1
  read $ foldl (\acc x -> acc ++ show x) "" wholeValues2 :: Integer

turnUltimateDialLeftUntilTheMessageAppears :: Integer -> Integer -> String
turnUltimateDialLeftUntilTheMessageAppears rotationsLeft smallerNumber = do
  let anIdiotsLuggageCombination = rotationsLeft
  let textNumber = show smallerNumber
  let order = [x * 4 | x <- [0..(length textNumber - 1)], x * 4 < (length textNumber - 1)]
  let y = [(read (take 4 (drop x textNumber)) :: Integer) - anIdiotsLuggageCombination | x <- order]
  let z = [ fromInteger x `shiftR` bitesShifted | x <- y]
  [if x == courseCorrection then chr (x `div` bassDecimal) else chr x | x <- z]



{--------------------------------

ASCII 3D Art

----------------------------------}

duplicateChar :: Char -> Int -> String
duplicateChar c x = [c | _<- [0..x]]

_createASCIIBoxWalls :: Int -> String
_createASCIIBoxWalls x = duplicateChar ' ' 4 ++ "|" ++ duplicateChar ' ' x ++ "|"

_createASCIIBoxBottom :: Int -> String
_createASCIIBoxBottom x = duplicateChar ' ' 5 ++ duplicateChar '-' x

_createASCIITopDepth :: Int -> Int -> Int -> String
_createASCIITopDepth offset x z = duplicateChar ' ' (5 + offset) ++ "/" ++ duplicateChar ' ' x ++ "/" ++ duplicateChar ' ' (z - offset - 2) ++ "|"

_createASCIIXTop :: Int -> Int -> Int -> String
_createASCIIXTop offset x z = duplicateChar ' ' (5 + offset) ++ "/" ++ duplicateChar '_' x ++ "/" ++ duplicateChar ' ' (z - offset - 2) ++ "|"

_createASCIIBoxTop :: Int -> String
_createASCIIBoxTop x = duplicateChar ' ' 5 ++ duplicateChar '_' x

createASCIIBox :: Int -> Int -> Int -> [String]
createASCIIBox xWidth yHeight zDepth
  | xWidth < 2 = error "Must have a width larger than 2"
  | yHeight < 2 = error "Must have a height larger than 2"
  | zDepth  < 1 = error "Must have a depth larger than 1"
  | zDepth > yHeight = error "zDepth cannot be larger than yHeight"
  | otherwise = do
    let realXWidth = xWidth * 3
    [duplicateChar ' ' (zDepth) ++ _createASCIIBoxTop realXWidth] ++
      [_createASCIITopDepth x realXWidth zDepth | x <- reverse [1..zDepth - 1]] ++
      [_createASCIIXTop 0 realXWidth (zDepth)] ++
      [] ++ (if yHeight - zDepth == 0 then [] else [_createASCIIBoxWalls realXWidth ++ duplicateChar ' ' (zDepth - 1) ++ "|" | _ <- reverse [0..(yHeight - zDepth)]]) ++
      [_createASCIIBoxWalls realXWidth ++ duplicateChar ' ' x ++ "/" | x <- reverse [0..zDepth - 1]] ++
      [_createASCIIBoxBottom realXWidth]


javascript/musks-ultimate-message-safe.js
/*
 Copyright (C) 2018 Michael Cornwell

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var pako = require('pako');
var bigInt = require('big-integer');

module.exports = class MusksUltimateMessageSafe {
  constructor () {
    this.BITES_SHIFTED = 0x04;
    this.COURSE_CORRECTION = 320;
    this.ULTIMATE_ANSWER = 42;
    this.BASS_DECIMAL = 10;
    this.OVER_HEXED = 0xF + 0x01;
    this.ULTIMATE_WINDAGE_ADJUSTMENT = this.ULTIMATE_ANSWER + 32;
    this.SECRET_SEED = new Uint8Array([
      0x78, 0x9c, 0x0d, 0x8e, 0x81, 0x09, 0x00, 0x40, 0x08, 0x02, 0x57,
      0xb2, 0xcf, 0xb4, 0xf6, 0x5f, 0xec, 0x45, 0x48, 0x0c, 0x2e, 0x7b,
      0x4f, 0xa5, 0xb3, 0x01, 0x3d, 0xf1, 0xe4, 0x3a, 0x0f, 0x6e, 0x20,
      0x6a, 0xb0, 0x3b, 0x15, 0x73, 0xd3, 0xbd, 0x99, 0x12, 0xad, 0x8b,
      0xde, 0xf1, 0x11, 0x0f, 0xdb, 0xa5, 0xa5, 0xec, 0x80, 0x53, 0x2f,
      0x11, 0xf1, 0x1b, 0x6f, 0xb8, 0xed, 0x7e, 0xe5, 0x67, 0x6e, 0x73,
      0x6e, 0x57, 0xc5, 0x54, 0x9d, 0xd0, 0x26, 0x3a, 0x72, 0x71, 0x72,
      0x60, 0xc8, 0x55, 0x36, 0xe1, 0xab, 0xaf, 0x76, 0x8d, 0x3c, 0xa2,
      0x71, 0x8f, 0x90, 0x26, 0x4c, 0xeb, 0x1a, 0x1f, 0xd2, 0x9b, 0x23,
      0x21]);
    this.TWENTY_TWO_SIDED_DIE_ENCASED_IN_WATER =
      [0, 8, 10, 12, 14, 16, 20, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 36, 38, 40];
    this.INDEX_TO_LETTER = {};
    this.LETTER_TO_INDEX = {};
    this.alphabet = "abcdefghijklmnopqrstuvwxyz";
    for(var i = 0; i < this.alphabet.length; ++i) {
      this.INDEX_TO_LETTER[i] = this.alphabet[i];
      this.LETTER_TO_INDEX[this.alphabet[i]] = i;
    }

    this.LUCKY_NUMBER = 0x03;
    this.HIGH = 0x01;


  }

  turnUltimateDialLeftToGetABigNumber(starting_number, rotations_left) {
    var number = bigInt(starting_number);
    number += rotations_left;
    number *= rotations_left;
    number = number ** this.ULTIMATE_ANSWER;
    number += bigInt(this._decryptSecretSeed());
    return number;
  }

  turnUltimateDialRightToGetASmallNumber(rotations_right, big_number) {
    big_number = String(big_number);
    var scrunched = big_number.slice(0, this.OVER_HEXED);
    var packed = big_number.slice(this.OVER_HEXED);
    var packed1 = packed.slice(0, this.ULTIMATE_WINDAGE_ADJUSTMENT);
    var packed2 = packed.slice(this.ULTIMATE_WINDAGE_ADJUSTMENT);
    var packed1Letters = this._convertNumberToLetters(packed1);
    var packed2Letters = this._convertNumberToLetters(packed2);
    var offset1 = this._unpackToOffsetList(packed1Letters);
    var offset2 = this._unpackToOffsetList(packed2Letters);
    var wholeValues = this._unscrunchBinary(scrunched);
    wholeValues = this._continueTurningLeftUntilClickIsHeard(wholeValues, offset2);
    wholeValues = this._continueTurningLeftUntilClickIsHeard(wholeValues, offset1);
    var singleValues = "";
    for (var val of wholeValues) {
      singleValues += String(val);
    }

    return singleValues;
  }

  turnUltimateDialLeftUntilTheMessageAppears(rotations_left, smaller_number) {
    var message = '';
    var an_idiots_luggage_combination = rotations_left;
    var count = 0;
    var smaller_number_text = String(smaller_number);

    while (count < smaller_number_text.length) {
      var number = parseInt(smaller_number_text.slice(count, count + 4));
      number -= an_idiots_luggage_combination;
      var shift = number >> this.BITES_SHIFTED;
      if (shift === this.COURSE_CORRECTION) {
        shift = parseInt(shift / this.BASS_DECIMAL);
      }

      message += String.fromCharCode(shift);
      count += 4;
    }

    return message;
  }

  _decryptSecretSeed() {
    var inflator = new pako.Inflate();
    var result = pako.inflate(this.SECRET_SEED);
    var result_string = "";
    for(var i = 0; i < result.length; ++i) {
      result_string += String.fromCharCode(result[i]);
    }

    return result_string;
  }

  _convertNumberToLetters(number) {
    number = String(number);
    var as_letters = "";
    var skip_next = false;
    for (var i = 0; i < number.length; ++i) {
      if (skip_next) {
        skip_next = false;
        continue;
      }

      var num = number[i];
      var as_int = parseInt(num);
      if (as_int === 1 || as_int === 2) {
        if (i === number.length - 1) {
          continue;
        }
        num += number[i + 1];
        let val = this._consultBlackBallWithNumber(parseInt(num));
        as_letters += val;
        skip_next = true;
      } else {
        let val = this._consultBlackBallWithNumber(as_int);
        as_letters += val;
      }
    }

    return as_letters;
  }

  _consultBlackBallWithNumber(value) {
    var size_of_twenty_two_sided_die = this.TWENTY_TWO_SIDED_DIE_ENCASED_IN_WATER.length;
    if (value === size_of_twenty_two_sided_die) {
      return this.INDEX_TO_LETTER[1];
    } else if (value === size_of_twenty_two_sided_die + 1) {
      return this.INDEX_TO_LETTER[2];
    } else if (value >= Object.keys(this.INDEX_TO_LETTER).length){
      return this.INDEX_TO_LETTER[0];
    }

    return this.INDEX_TO_LETTER[value];
  }

  _consultBlackBallWithLetter(letter) {
      var value = 0;
      for (var i = 0; i < this.alphabet.length; ++i) {
        if (letter === this.alphabet[i]) {
          value = this.TWENTY_TWO_SIDED_DIE_ENCASED_IN_WATER[i];
          if (value === undefined) {
            value = 0;
          }

          break;
        }
      }

      return value;
    }

  _unpackToOffsetList(letters) {
    var final = [];
    for (var l of letters) {
      var result = this._consultBlackBallWithLetter(l);
      final.push(result);
    }

    return final;
  }

  _unscrunchBinary(scrunched) {
    var results = [];
    scrunched = bigInt(scrunched);
    while (scrunched > 1) {
      if ((scrunched & 0b1)) {
        results.unshift(this.LUCKY_NUMBER);
      } else {
        results.unshift(this.HIGH);
      }

      scrunched = scrunched.shiftRight(1);
    }
    return results;
  }

  _continueTurningLeftUntilClickIsHeard(whole_values, offsets) {
    var as_ints = [];
    for (var i = 0; i < whole_values.length; ++i) {
      var value = whole_values[i];
      var offset = offsets[i];
      as_ints.push((value * this.ULTIMATE_ANSWER) + offset);
    }

    return as_ints
  }






};

